### PR TITLE
feat: allow adding/removing tags from DS docs (core)

### PR DESCRIPTION
**Pull request by fontanierh**
adds a new endpoint allowing to update a `data_sources_document`'s tags (specifying `add_tags` and `remove_tags`)

to keep in mind:
- QDrant update happens after DB update, so technically if QDrant is down we'll end up out of sync
- `add_tags` will add the tags to the existing ones, while `remove_tags` will remove those tags from the exisitng ones. There's no "replace_tags"
- I need this to implement the `__DUST_TRACK` (system) tag for the document_tracker exploration

### CODE REVIEW

File: core/bin/dust_api.rs [hunk 1]
```diff
@@ -1156,6 +1157,126 @@ async fn data_sources_search(
...
+    let add_tags = match payload.add_tags {
+        Some(tags) => tags,
+        None => vec![],
+    };
+    let remove_tags = match payload.remove_tags {
+        Some(tags) => tags,
+        None => vec![],
+    };
+    let add_tags_set: HashSet<String> = add_tags.iter().cloned().collect();
+    if add_tags.len() != add_tags_set.len() {
```
Comment (spolu) [inline core/bin/dust_api.rs:1183]: not sure these errors are really needed. I would remove them

File: core/bin/dust_api.rs [hunk 2]
```diff
@@ -1156,6 +1157,126 @@ async fn data_sources_search(
...
+        );
+    }
+    if add_tags_set.intersection(&remove_tags_set).count() > 0 {
+        return (
+            StatusCode::BAD_REQUEST,
+            Json(APIResponse {
+                error: Some(APIError {
+                    code: String::from("bad_request"),
+                    message: String::from(
+                        "The `add_tags` and `remove_tags` parameters contain duplicates.",
```
Comment (spolu) [inline core/bin/dust_api.rs:1215]: not contains duplicate but rather have a non empty intersection

File: core/src/data_sources/data_source.rs [hunk 3]
```diff
@@ -338,6 +339,50 @@ impl DataSource {
...
+            ..Default::default()
+        };
+        let points_selector = PointsSelector {
+            points_selector_one_of: Some(PointsSelectorOneOf::Filter(Filter {
+                must: vec![field_condition.into()],
+                ..Default::default()
+            })),
+        };
+        qdrant_client
+            .set_payload(
```
Comment (spolu) [inline core/src/data_sources/data_source.rs:375]: We're sure it does not erase the rest of the payload?

**Review by spolu — DISMISSED** [review]
LGTM

Left a question and a micro nit.

File: core/src/data_sources/data_source.rs (hunk 3 above)
Comment (spolu) [reply core/src/data_sources/data_source.rs:375]: cc @philipperolet who I believe played with that part of qdrant during the timetamp migation. Can you take a look as well? :pray:

Comment (fontanierh) [reply core/src/data_sources/data_source.rs:375]: Yes ! I tested that

File: core/bin/dust_api.rs (hunk 1 above)
Comment (fontanierh) [reply core/bin/dust_api.rs:1183]: Ok for me. Should I still dedup though ?

File: core/bin/dust_api.rs (hunk 2 above)
Comment (fontanierh) [reply core/bin/dust_api.rs:1215]: Aligned, updating the wording 👍

File: core/bin/dust_api.rs (hunk 1 above)
Comment (fontanierh) [reply core/bin/dust_api.rs]: kept the de-dup (rust's `.collect()` is awesome)

File: core/bin/dust_api.rs (hunk 2 above)
Comment (fontanierh) [reply core/bin/dust_api.rs]: fixed

File: core/bin/dust_api.rs (hunk 1 above)
Comment (spolu) [reply core/bin/dust_api.rs]: yes yes dedup but not fail :+1:

**Review by spolu — APPROVED** [review]
LGTM

Would love to have final :+1: from @philipperolet

**Review by philipperolet — APPROVED** [review]
yes I used set_payload for the timetttttttttamp migration; it does not erase the rest of the payload

LGTM -- although did not look in detail at the rust part, not yet Rusty enough :)

**Comment by fontanierh** [comment]
Thanks 🙌

**Pull request merged**